modules/tts/tts-utils/tts-adapter-google/src/main/java/org/daisy/pipeline/tts/google/impl/GoogleRestTTSEngine.java

```diff
@@ -0,0 +1,262 @@
+		mAudioFormat = audioFormat;
+		mRequestScheduler = requestScheduler;
+		mRequestBuilder = new GoogleRequestBuilder(apiKey);
+		
+	}
+
+	@Override
+	public Collection<AudioBuffer> synthesize(String sentence, XdmNode xmlSentence,
+			Voice voice, TTSResource threadResources, AudioBufferAllocator bufferAllocator, boolean retry)
+					throws SynthesisException,InterruptedException, MemoryException {
+		
+		if (sentence.length() > 5000) {
+			throw new SynthesisException("The number of characters in the sentence must not exceed 5000.");
+		}
+
+		Collection<AudioBuffer> result = new ArrayList<AudioBuffer>();
+
+		// the sentence must be in an appropriate format to be inserted in the json query
+		// it is necessary to wrap the sentence in quotes and add backslash in front of the existing quotes
+		
+		String adaptedSentence = "";
+		
+		for (int i = 0; i < sentence.length(); i++) {
+			if (sentence.charAt(i) == '"') {
+				adaptedSentence = adaptedSentence + '\\' + sentence.charAt(i);
+			}
+			else {
+				adaptedSentence = adaptedSentence + sentence.charAt(i);
+			}
+		}
+		
+		String languageCode;
+		String name;
+		int indexOfSecondHyphen;
+
+		if (voice != null) {
+			//recovery of the language code in the name of the voice
+			indexOfSecondHyphen = voice.name.indexOf('-', voice.name.indexOf('-') + 1);
+			languageCode = voice.name.substring(0, indexOfSecondHyphen);
+			name = voice.name;
+		}
+		else {
+			// by default the voice is set to English
+			languageCode = "en-GB";
+			name = "en-GB-Standard-A";
+		}
+		
+		GoogleRestRequest speechRequest = null;
+		GoogleRestRequest request = null;
+		UUID requestUuid = null;
+		boolean isNotDone = true;
+		
+		try {
+			
+			speechRequest = mRequestBuilder.newRequest()
+					.withSampleRate((int)mAudioFormat.getSampleRate())
+					.withAction(GoogleRestAction.SPEECH)
+					.withLanguageCode(languageCode)
+					.withVoice(name)
+					.withText(adaptedSentence)
+					.build();
+			
+			requestUuid = mRequestScheduler.add(speechRequest);
+			
+		} catch (Throwable e) {
+			throw new SynthesisException(e.getMessage(), e.getCause());
+		}
+		
+		// we loop until the request has not been processed 
+		// (google limits to 300 requests per minute or 15000 characters)
+		while(isNotDone) {
+			
+			try {
+				
+				request = mRequestScheduler.poll(requestUuid);
+
+				BufferedReader br = new BufferedReader(new InputStreamReader(request.send(), "utf-8"));
+				StringBuilder response = new StringBuilder();
+				String inputLine;
+				while ((inputLine = br.readLine()) != null) {
+					response.append(inputLine.trim());
+				}
+				br.close();
+
+				// the answer is encoded in base 64, so it must be decoded
+				byte[] decodedBytes = Base64.getDecoder().decode(response.toString().substring(18, response.length()-2));
+
+				AudioBuffer b = bufferAllocator.allocateBuffer(decodedBytes.length);
+				b.data = decodedBytes;
+				result.add(b);
+				
+				isNotDone = false;
+
+			} catch (Throwable e1) {
```

Comment:
I find this not so great. Isn't there a more specific exception we can catch when we get a 429 error?

Reply:
it is the more common error code expected (quotas are exceeded).
i'll have to check the doc of google for the other possible code.

Reply:
I mean, this code expects that when you get a 429, an exception is thrown somewhere in the try block. So why not catch specifically that exception?

Reply:
haaa sorry misunderstood.
I think it is due to the scheduler interface that can pause the thread (and so can throw an InterruptedException that needs to be catched).
I definitely need to rework the scheduler interface, i'm working on a proposal based on your previous comment to export more of the logic for the different rest engines to the tts-common.